sgkit/stats/aggregation.py

```diff
@@ -1,10 +1,103 @@
 import dask.array as da
+import numba
 import numpy as np
 import xarray as xr
 from xarray import DataArray, Dataset
 
+from ..typing import ArrayLike
 
-def count_alleles(ds: Dataset) -> DataArray:
+
+@numba.guvectorize(  # type: ignore
+    [
+        "void(numba.int8[:], numba.uint8[:], numba.uint8[:])",
+        "void(numba.int16[:], numba.uint8[:], numba.uint8[:])",
+        "void(numba.int32[:], numba.uint8[:], numba.uint8[:])",
+        "void(numba.int64[:], numba.uint8[:], numba.uint8[:])",
+    ],
+    "(k),(n)->(n)",
+    nopython=True,
+)
+def count_alleles(g: ArrayLike, _: ArrayLike, out: ArrayLike) -> None:
+    """Generaliszed U-function for computing per sample allele counts.
+
+    Parameters
+    ----------
+    g : (K,) array-like, int
+        A genotype call with K alleles where K is the genotypes ploidy.
+    _: (N,) array-like, uint8
+        Dummy variable of length N where N is the number of possible
+        unique alleles.
+
+    Returns
+    -------
+    ac : (N,) array-like, uint8
+        Allele counts with values corresponding to the number of
+        non-missing occurrences of each allele whithin g.
+
+    """
+    out[:] = 0
+    n_allele = len(g)
+    for i in range(n_allele):
+        a = g[i]
+        if a >= 0:
+            out[a] += 1
+
+
+def count_call_alleles(ds: Dataset) -> DataArray:
+    """Compute per sample allele counts from genotype calls.
+
+    Parameters
+    ----------
+    ds : Dataset
+        Genotype call dataset such as from
+        `sgkit.create_genotype_call_dataset`.
+
+    Returns
+    -------
+    call_allele_count : DataArray
+        Allele counts with shape (variants, samples, alleles) and values
+        corresponding to the number of non-missing occurrences
+        of each allele.
+
+    Examples
+    --------
+
+    >>> import sgkit as sg
+    >>> from sgkit.testing import simulate_genotype_call_dataset
+    >>> ds = simulate_genotype_call_dataset(n_variant=4, n_sample=2, seed=1)
+    >>> ds['call_genotype'].to_series().unstack().astype(str).apply('/'.join, axis=1).unstack() # doctest: +NORMALIZE_WHITESPACE
```

Comment:
This is a good place for @tomwhite's code in https://github.com/pystatgen/sgkit/pull/58 now.